#34234 normalize(a) for multidimensional arrays

ssikdar1 opened the pull request:
PR to try and resolve issue JuliaLang/LinearAlgebra.jl#688

- [X]  Change the type signatures and documentation from AbstractVector to AbstractArray
- [x] Change this line from v[1] to first(v). (This is a bug anyway, since v[1] is wrong for non 1-based arrays.)
- [X] Change the documentation to refer to "array" rather than "vector" and use a rather than v. (changed docstrings)
- [X] Add a few tests. (not sure if i need more?)
- [X] Add a NEWS item.

Testing: 
added unit test to ` stdlib/LinearAlgebra/test/generic.jl`
ran ` ./julia stdlib/LinearAlgebra/test/generic.jl` with no errors.

```
...
Test Summary:                         | Pass  Total
normalize for multidimensional arrays |    2      2
...
```

Testing in the shell:

```
julia> using LinearAlgebra;

julia> arr = [ [1.0 0.0]; [0.0 1.0] ]
2×2 Array{Float64,2}:
 1.0  0.0
 0.0  1.0

julia> @show arr
arr = [1.0 0.0; 0.0 1.0]
2×2 Array{Float64,2}:
 1.0  0.0
 0.0  1.0

julia> @show normalize(arr)
normalize(arr) = [0.7071067811865475 0.0; 0.0 0.7071067811865475]
2×2 Array{Float64,2}:
 0.707107  0.0
 0.0       0.707107

julia> @show normalize!(copy(arr))
normalize!(copy(arr)) = [0.7071067811865475 0.0; 0.0 0.7071067811865475]
2×2 Array{Float64,2}:
 0.707107  0.0
 0.0       0.707107

julia> @show normalize(arr) == normalize!(copy(arr))
normalize(arr) == normalize!(copy(arr)) = true
true

julia> arr = [[1.0 0.0 0.0]; [0.0 1.0 0.0]]
2×3 Array{Float64,2}:
 1.0  0.0  0.0
 0.0  1.0  0.0

julia> @show normalize(arr) == normalize!(copy(arr))
normalize(arr) == normalize!(copy(arr)) = true
true

julia> a = [1,0,0]
3-element Array{Int64,1}:
 1
 0
 0

julia> @show normalize(a) == normalize!(copy(a))
normalize(a) == normalize!(copy(a)) = true
true
```

ssikdar1 committed:
Add support normalize multi dim arrays

ssikdar1 committed:
remove trailing whitespace from test

ssikdar1 committed:
var name v => a for inner function

oscardssmith commented:
Should this be backported? It seems simple and is partially a bug fix

StefanKarpinski commented:
No, this is definitely a feature.

jw3126 commented on stdlib/LinearAlgebra/test/generic.jl:
```diff
@@ -248,6 +248,18 @@ end
 end
 
+@testset "normalize for multidimensional arrays" begin
+    arr = [ [1.0 0.0]; [0.0 1.0] ]
```
This looks a bit like numpy but it is not how you do it in julia. You can simply create a `Matrix` like e.g. `[a b; c d]`.

ssikdar1 replied:
updated in f602363

jw3126 commented on stdlib/LinearAlgebra/test/generic.jl:
```diff
@@ -248,6 +248,18 @@ end
+            [0.0 1.0 0.0]
+         ]
+    @test normalize(arr) == normalize!(copy(arr))
+
```
It would be good to test that this is consistent with the vector case. Some tests like
```julia
arr = randn(2,3)
@test size(normalize(arr)) == size(arr)
@test vec(normalize(arr)) == normalize(vec(arr))
```

ssikdar1 replied:
added a test case based on a testcase I found in test/offset.jl

```
julia> A = OffsetArray([-1,0], (-2,))
2-element OffsetArray{Int64,1,Array{Int64,1}} with indices -1:0:
 -1
  0
julia> Arsc = reshape(A, :, 1)
2×1 reshape(::OffsetArray{Int64,1,Array{Int64,1}}, 2, 1) with eltype Int64:
 -1
  0
julia> Arsc[1,1] = 5
5
julia> first(A)
5
julia> A[1]
ERROR: BoundsError: attempt to access 2-element OffsetArray{Int64,1,Array{Int64,1}} with indices -1:0 at index [1]
```
Works with normalize now:
```
julia> normalize(A)
2-element OffsetArray{Float64,1,Array{Float64,1}} with indices -1:0:
 1.0
 0.0
```

jw3126 commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1638,15 +1638,29 @@ julia> c = normalize(a, 1)
+function normalize(a::AbstractArray, p::Real = 2)
+    nrm = norm(a, p)
+    if !isempty(a)
+        aa = copy_oftype(a, typeof(first(a)/nrm))
```
Cool that you found this bug! Can you add a test with a e.g. OffaxisArray that does not have index 1?

ssikdar1 replied:
added a test case based on a testcase I found in test/offset.jl

```
julia> A = OffsetArray([-1,0], (-2,))
2-element OffsetArray{Int64,1,Array{Int64,1}} with indices -1:0:
 -1
  0
julia> Arsc = reshape(A, :, 1)
2×1 reshape(::OffsetArray{Int64,1,Array{Int64,1}}, 2, 1) with eltype Int64:
 -1
  0
julia> Arsc[1,1] = 5
5
julia> first(A)
5
julia> A[1]
ERROR: BoundsError: attempt to access 2-element OffsetArray{Int64,1,Array{Int64,1}} with indices -1:0 at index [1]
```
Works with normalize now:
```
julia> normalize(A)
2-element OffsetArray{Float64,1,Array{Float64,1}} with indices -1:0:
 1.0
 0.0
```

jw3126 replied:
Yeah that is a good test case!

ssikdar1 committed:
Update normalize tests

Case for OffsetArray where A[1] would fail but
first(A) would not. Also some more test cases to
compare with the vector case

ssikdar1 committed:
add NEWS item

ssikdar1 commented:
The CI failures seem to be complaining about the Makefile and seem to be unrelated?
```
make: *** No rule to make target 'win-extras'.  Stop.
```

jw3126 commented on stdlib/LinearAlgebra/src/generic.jl:
```diff
@@ -1638,15 +1638,29 @@ julia> c = normalize(a, 1)
 julia> norm(c, 1)
 1.0
+
+julia> a = [[1 2 4] ; [1 2 4]]
```
```suggestion
julia> a = [1 2 4 ; 1 2 4]
```
same below.

ssikdar1 replied:
addressed in  deaf6b4

stevengj commented:
LGTM.  Thanks for tackling this, @ssikdar1!

ssikdar1 committed:
make docstring example w/ array  more julia-thonic

ssikdar1 committed:
Merge remote-tracking branch 'upstream/master'

stevengj commented on NEWS.md:
```diff
@@ -30,8 +30,8 @@ Standard library changes
 #### LinearAlgebra
-
 * The BLAS submodule now supports the level-2 BLAS subroutine `hpmv!` ([#34211]).
+* `normalize` now supports multidimensional arrays ([#34234])
```
should be 34239 (the PR, not the issue)

ssikdar1 replied:
sorry! fixed in  cdb2167

stevengj commented on stdlib/LinearAlgebra/test/generic.jl:
```diff
@@ -248,6 +253,31 @@ end
+    arr = [
+            [1.0 0.0 0.0];
+            [0.0 1.0 0.0]
+         ]
```
`arr = [ 1.0 0.0 0.0; 0.0 1.0 0.0 ]` or
```
arr = [ 1.0 0.0 0.0
            0.0 1.0 0.0 ]
````

ssikdar1 replied:
removed test case in  cdb2167

stevengj commented on stdlib/LinearAlgebra/test/generic.jl:
```diff
@@ -248,6 +253,31 @@ end
+    test_arr(arr)
+
+    arr = randn(2,3)
+    test_arr(arr)
```
I don't really see the point of having three test cases for 2d arrays here.  Seems like
```
test_arr([1 2 3; 4 5 6])
test_arr(OffsetArray([-1,0], (-2,)))
```
would be sufficient to test all of the cases.

In such situations, rather than define a `test_arr` function, it is more idiomatic to use a loop:
```
for A in ([1 2 3; 4 5 6], OffsetArray([-1,0], (-2,)))
    @test normalize(arr) == normalize!(copy(arr))
    @test size(normalize(arr)) == size(arr)
    @test vec(normalize(arr)) == normalize(vec(arr))
end
```

ssikdar1 replied:
I removed the redundant test cases in cdb2167.

An interesting thing popped up:

```
 using LinearAlgebra;
 
arr = [ 1 2 3; 4 5 6 ] 
normalize!(copy(arr))
```
Actually throws an error somewhere in the trace stack its trying to do a Int64 cast of a decimal:

```
$ ./julia foo.jl 
ERROR: LoadError: InexactError: Int64(0.10482848367219183)
Stacktrace:
 [1] Int64 at ./float.jl:710 [inlined]
 [2] convert at ./number.jl:7 [inlined]
 [3] setindex! at ./array.jl:825 [inlined]
 [4] macro expansion at /Users/ssikdar1/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/generic.jl:150 [inlined]
 [5] macro expansion at ./simdloop.jl:77 [inlined]
 [6] rmul!(::Array{Int64,2}, ::Float64) at /Users/ssikdar1/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/generic.jl:149
 [7] __normalize! at /Users/ssikdar1/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/generic.jl:1602 [inlined]
 [8] normalize!(::Array{Int64,2}, ::Int64) at /Users/ssikdar1/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/generic.jl:1593
 [9] normalize!(::Array{Int64,2}) at /Users/ssikdar1/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/generic.jl:1592
 [10] top-level scope at /Users/ssikdar1/julia/foo.jl:5
 [11] include(::Module, ::String) at ./Base.jl:377
 [12] exec_options(::Base.JLOptions) at ./client.jl:288
 [13] _start() at ./client.jl:484
in expression starting at /Users/ssikdar1/julia/foo.jl:5
``` 

When the array is changed to float numbers:
```
using LinearAlgebra;
   
arr = [ 1.0 2.0 3.0; 4.0 5.0 6.0 ]
normalize!(copy(arr))
```
the code works.

Should the cast issue be resolved in this PR or should I open a new github issue?

jw3126 replied:
If you run `@show typeof([ 1 2 3; 4 5 6 ] )` you will see it is an array of `Int`. 
`normalize!` tries to mutate the array inplace. This cannot change the eltype of the array, it will still be an array of `Int`. But it is very hard for an array of `Int` to have norm one (it means there is a single nonzero entry that is +-1). So I would say it is expected that `normalize!` almost always throws an error on `Int` arrays. However `normalize` without `!` should work on `Int` arrays and return `Float64` arrays. This should be tested.

ssikdar1 committed:
reduce redundant test cases

ssikdar1 committed:
add test for normalize on Int64 array

jw3126 reviewed (approved):
LGTM, thanks very much @ssikdar1 for doing this!

jw3126 commented on stdlib/LinearAlgebra/test/generic.jl:
```diff
@@ -248,6 +253,18 @@ end
 
+@testset "normalize for multidimensional arrays" begin
+
+    for arr in ([1.0 2.0 3.0; 4.0 5.0 6.0], OffsetArray([-1,0], (-2,)))
```
Adding a few higher dimensional arrays can't hurt:
```suggestion
    for arr in ([1.0 2.0 3.0; 4.0 5.0 6.0], OffsetArray([-1,0], (-2,)), rand(1,2,3), rand(1,2,3,4))
```

jw3126 replied:
Okay this is super nitpicky, but we should also test the zero dimensional case. You can create zero dimensional arrays like:
```julia
julia> fill(10.0, ())
0-dimensional Array{Float64,0}:
10.0
```

ssikdar1 replied:
ok! with ae84cb3 there are now Array{Float64,0}, Array{Float64,1}, Array{Float64,3}, Array{Float64,3}, Array{Float64,4} test cases as well as the offset array and the Int64 case.

jw3126 replied:
Thanks again @ssikdar1 for doing this and all your patience for pedantic requests.

ssikdar1 replied:
yup! no worries!

ssikdar1 committed:
add 0 1 and high dim test cases
